File: ios/Runner/FlutterDependencyProvider.swift
import Foundation
import UIKit
import LibSignalClient

/// Dependency provider for Flutter app's native Signal Protocol components on iOS.
/// Similar to FlutterDependencyProvider on Android but adapted for iOS/Swift.
class FlutterDependencyProvider {
    private static var cryptographicService: SekretessCryptographicService?
    private static var signalProtocolStore: SekretessSignalProtocolStore?
    private static var apiCallback: ((String, [String: Any]?) -> Bool)?
    
    static func initialize() {
        // Initialize database
        let database = SekretessDatabase.shared
        
        // Initialize repositories
        let identityKeyRepository = IdentityKeyRepository(database: database)
        let registrationRepository = RegistrationRepository(database: database)
        let preKeyRepository = PreKeyRepository(database: database)
        let signedPreKeyRepository = SignedPreKeyRepository(database: database)
        let sessionRepository = SessionRepository(database: database)
        let senderKeyRepository = SenderKeyRepository(database: database)
        let kyberPreKeyRepository = KyberPreKeyRepository(database: database)
        
        // Initialize Signal Protocol Store
        signalProtocolStore = SekretessSignalProtocolStore(
            identityKeyRepository: identityKeyRepository,
            registrationRepository: registrationRepository,
            preKeyRepository: preKeyRepository,
            signedPreKeyRepository: signedPreKeyRepository,
            sessionRepository: sessionRepository,
            senderKeyRepository: senderKeyRepository,
            kyberPreKeyRepository: kyberPreKeyRepository
        )
        
        // Initialize Cryptographic Service
        guard let store = signalProtocolStore else {
            fatalError("Failed to initialize Signal Protocol Store")
        }
        
        cryptographicService = SekretessCryptographicService(signalProtocolStore: store)
        
        print("FlutterDependencyProvider initialized")
    }
    
    static func getCryptographicService() -> SekretessCryptographicService {
        guard let service = cryptographicService else {
            fatalError("FlutterDependencyProvider not initialized. Call initialize() first.")
        }
        return service
    }
    
    static func setApiCallback(_ callback: @escaping (String, [String: Any]?) -> Bool) {
        apiCallback = callback
    }
    
    static func callApi(method: String, arguments: [String: Any]?) -> Bool {
        guard let callback = apiCallback else {
            print("API callback not set")
            return false
        }
        return callback(method, arguments)
    }
    
    static func getApplicationContext() -> Any {
        // Return UIApplication for compatibility
        return UIApplication.shared
    }
}

File: android/app/src/main/java/io/sekretess/bridge/FlutterDependencyProvider.java
package io.sekretess.bridge;

import android.content.Context;
import android.util.Log;

import io.sekretess.cryptography.storage.SekretessSignalProtocolStore;
import io.sekretess.db.SekretessDatabase;
import io.sekretess.db.repository.IdentityKeyRepository;
import io.sekretess.db.repository.KyberPreKeyRepository;
import io.sekretess.db.repository.PreKeyRepository;
import io.sekretess.db.repository.RegistrationRepository;
import io.sekretess.db.repository.SenderKeyRepository;
import io.sekretess.db.repository.SessionRepository;
import io.sekretess.db.repository.SignedPreKeyRepository;
import io.sekretess.service.SekretessCryptographicService;

/**
 * Dependency provider for Flutter app's native Signal Protocol components.
 * Similar to SekretessDependencyProvider but adapted for Flutter.
 */
public class FlutterDependencyProvider {
    private static final String TAG = "FlutterDependencyProvider";
    
    private static SekretessCryptographicService cryptographicService;
    private static SekretessSignalProtocolStore signalProtocolStore;
    private static NativeApiClientBridge apiClientBridge;
    private static Context applicationContext;
    
    public static void initialize(Context context) {
        // Prevent double initialization
        if (applicationContext != null && cryptographicService != null) {
            Log.i(TAG, "FlutterDependencyProvider already initialized, skipping");
            return;
        }
        
        applicationContext = context.getApplicationContext();
        
        // Initialize database
        SekretessDatabase database = SekretessDatabase.getInstance(applicationContext);
        
        // Initialize repositories (they will get database instance internally)
        IdentityKeyRepository identityKeyRepository = new IdentityKeyRepository();
        RegistrationRepository registrationRepository = new RegistrationRepository();
        PreKeyRepository preKeyRepository = new PreKeyRepository();
        SignedPreKeyRepository signedPreKeyRepository = new SignedPreKeyRepository();
        SessionRepository sessionRepository = new SessionRepository();
        SenderKeyRepository senderKeyRepository = new SenderKeyRepository();
        KyberPreKeyRepository kyberPreKeyRepository = new KyberPreKeyRepository();
        
        // Initialize Signal Protocol Store
        signalProtocolStore = new SekretessSignalProtocolStore(
            identityKeyRepository,
            registrationRepository,
            preKeyRepository,
            signedPreKeyRepository,
            sessionRepository,
            senderKeyRepository,
            kyberPreKeyRepository
        );
        
        // Initialize API client bridge
        apiClientBridge = new NativeApiClientBridge(applicationContext);
        
        // Initialize Cryptographic Service
        cryptographicService = new SekretessCryptographicService(signalProtocolStore);
        cryptographicService.setApiClientBridge(apiClientBridge);
        
        Log.i(TAG, "FlutterDependencyProvider initialized");
    }
    
    public static SekretessCryptographicService getCryptographicService() {
        if (cryptographicService == null) {
            throw new IllegalStateException("FlutterDependencyProvider not initialized. Call initialize() first.");
        }
        return cryptographicService;
    }
    
    public static NativeApiClientBridge getApiClientBridge() {
        if (apiClientBridge == null) {
            throw new IllegalStateException("FlutterDependencyProvider not initialized. Call initialize() first.");
        }
        return apiClientBridge;
    }
    
    public static Context getApplicationContext() {
        if (applicationContext == null) {
            throw new IllegalStateException("FlutterDependencyProvider not initialized. Call initialize() first.");
        }
        return applicationContext;
    }
    
    // Static method for compatibility with repository constructors
    public static Context applicationContext() {
        return getApplicationContext();
    }
}
